mlflow/store/sqlalchemy_store.py

```diff
@@ -0,0 +1,222 @@
+                            start_time=start_time, end_time=None,
+                            source_version=source_version,
+                            lifecycle_stage=entities.RunInfo.ACTIVE_LIFECYCLE)
+
+        for tag in tags:
+            run.tags.append(models.SqlRunTag(key=tag.key, value=tag.value))
+        self._save_to_db([run])
+
+        run = run.to_mlflow_entity()
+
+        return run
+
+    def update_run_info(self, run_uuid, run_status, end_time):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_uuid).first()
+        run.status = run_status
+        run.end_time = end_time
+
+        self._save_to_db(run)
+        run = run.to_mlflow_entity()
+
+        return run.info
+
+    def restore_run(self, run_id):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_id).first()
+        run.is_deleted = False
+        self._save_to_db(run)
+
+    def get_run(self, run_uuid):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_uuid,
+                                                          is_deleted=False).first()
+        if run is None:
+            raise MlflowException('Run(uuid={}) doesn\'t exist'.format(run_uuid),
+                                  error_codes.RESOURCE_DOES_NOT_EXIST)
+
+        return run.to_mlflow_entity()
+
+    def delete_run(self, run_id):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_id).first()
+        run.is_deleted = True
+        self._save_to_db(run)
+
+    def log_metric(self, run_uuid, metric):
+        try:
+            self._get_or_create(models.SqlMetric, run_uuid=run_uuid, key=metric.key,
+                                value=metric.value, timestamp=metric.timestamp)
+        except IntegrityError:
+            raise MlflowException('Metric={} must be unique'.format(metric),
+                                  error_codes.INVALID_PARAMETER_VALUE)
+
+    def log_param(self, run_uuid, param):
+        # if we try to update the value of an existing param this will fail
+        # because it will try to create it with same run_uuid, param key
+        try:
+            self._get_or_create(models.SqlParam, run_uuid=run_uuid, key=param.key,
+                                value=param.value)
+        except IntegrityError:
+            raise MlflowException('changing parameter {} value is not allowed'.format((run_uuid,
+                                                                                      param)),
+                                  error_codes.INVALID_PARAMETER_VALUE)
+
+    def set_tag(self, run_uuid, tag):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_uuid).first()
+
+        new_tag = models.SqlRunTag(key=tag.key, value=tag.value)
+        run.tags.append(new_tag)
+        self._save_to_db([run, new_tag])
+
+    def get_metric(self, run_uuid, metric_key):
+        run = self.get_run(run_uuid)
```

Comment:
for `get_tag`, `get_param`, `get_metric`, and `get_metric_history` -- would be more efficient to hit these databases. `get_run` would be expensive as it would hydrate all this data before return one of these.

in fact, `get_run` can instead depend on these methods.